File: crates/klyron_updater/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::UpdaterBuilder::Impl {
public:
  Klyron::UpdaterConfig config;
  bool verbose = false;
};

Klyron::UpdaterBuilder::Klyron::UpdaterBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::UpdaterBuilder& Klyron::UpdaterBuilder::withConfig(const Klyron::UpdaterConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::UpdaterBuilder& Klyron::UpdaterBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::UpdaterBuilder::Klyron::UpdaterInstance Klyron::UpdaterBuilder::build() {
  Klyron::UpdaterInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_updater/bindings/rust/builder.rs
//! Builder pattern for klyron_updater

use crate::types::Klyron::UpdaterConfig;

#[derive(Debug, Default)]
pub struct Klyron::UpdaterBuilder {
    config: Option<Klyron::UpdaterConfig>,
    verbose: bool,
}

impl Klyron::UpdaterBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::UpdaterConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::UpdaterInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::UpdaterInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::UpdaterInstance {
    pub config: Klyron::UpdaterConfig,
    pub verbose: bool,
}
